bokehjs/src/coffee/models/tools/draw/point_draw_tool.ts

```diff
@@ -0,0 +1,120 @@
+import * as p from "core/properties"
+import {ColumnDataSource} from "models/sources/column_data_source"
+import {SelectTool} from "../gestures/select_tool"
+import {EditToolView} from "./edit_tool"
+
+
+export interface BkEv {
+  bokeh: {
+    sx: number
+    sy: number
+  }
+  srcEvent: {
+    shiftKey?: boolean
+  }
+  keyCode: number
+  timeStamp: number
+}
+
+
+export class PointDrawView extends EditToolView {
+  model: PointDrawTool
+
+  _tap(e: BkEv) {
+    const ds = this.model.source;
+    const indices = ds.selected['1d'].indices.slice(0);
+    const append = e.srcEvent.shiftKey != null ? e.srcEvent.shiftKey : false;
+    const did_hit = this._select_event(e, append);
+    if (did_hit) {
+      return null;
+    }
+
+    const [x, y] = this._map_drag(e.bokeh.sx, e.bokeh.sy);
+    let xs = ds.data[this.model.x];
+    let ys = ds.data[this.model.y];
+    if ((xs.concat == null)) {
+      xs = Array.prototype.slice.call(xs);
+      ds.data[this.model.x] = xs;
+    }
+    if ((ys.concat == null)) {
+      ys = Array.prototype.slice.call(ys);
+      ds.data[this.model.y] = ys;
+    }
+    for (let k in ds.data) {
+      let v = ds.data[k];
+      if (![xs, ys].includes(v)) {
+        if ((v.push == null)) {
+          ds.data[k] = (v = Array.prototype.slice.call(v));
+        }
+        v.push('');
+      }
+    }
```

Comment:
What is going on here?  My wild guess is that this has something to do with typed arrays. Typing `xs` and `ys` would be a start to clean this up. Additionally, bokehjs' `includes()` should be used instead of `Array.includes()` (not available in IE). However, given that `v` is an array, then this code likely wouldn't work anyway, because JavaScript supports only reference equality for non-primitive objects. If you want universal equality you have to use bokehjs' `isEquals()`.

Reply:
Right, this code converts typed arrays to non-typed arrays, so that data can be added more easily added and removed. When inserting a point x, y value all other columns need to be filled with empty data to avoid the ColumnDataSource column length warnings and maintain consistency. Instead of an empty string I'll likely change the empty value to NaN though.

>JavaScript supports only reference equality for non-primitive objects

Reference equality is all that's needed but I'll make the equality checks explicit anyway.